javalib/src/main/scala/scala/scalanative/nio/fs/NativePosixFileAttributeView.scala

```diff
@@ -34,71 +33,120 @@ final class NativePosixFileAttributeView(path: Path,
       throw new IOException()
   }
 
-  override def setOwner(owner: UserPrincipal): Unit = {
-    val passwd = getPasswd(toCString(owner.getName))
-    if (unistd.chown(toCString(path.toString), !(passwd._2), -1.toUInt) != 0)
-      throw new IOException()
-  }
+  override def setOwner(owner: UserPrincipal): Unit =
+    Zone { implicit z =>
+      val passwd = getPasswd(toCString(owner.getName))
+      if (unistd.chown(toCString(path.toString), !(passwd._2), -1.toUInt) != 0)
+        throw new IOException()
+    }
 
-  override def setPermissions(perms: Set[PosixFilePermission]): Unit = {
-    var mask = 0.toUInt
-    NativePosixFileAttributeView.permMap.foreach {
-      case (flag, value) => if (perms.contains(value)) mask = mask | flag
+  override def setPermissions(perms: Set[PosixFilePermission]): Unit =
+    Zone { implicit z =>
+      var mask = 0.toUInt
+      NativePosixFileAttributeView.permMap.foreach {
+        case (flag, value) => if (perms.contains(value)) mask = mask | flag
+      }
+      if (stat.chmod(toCString(path.toString), mask) != 0) {
+        throw new IOException()
+      }
     }
-    if (stat.chmod(toCString(path.toString), mask) != 0)
-      throw new IOException()
-  }
 
   override def getOwner(): UserPrincipal =
     attributes.owner
 
-  override def setGroup(group: GroupPrincipal): Unit = {
-    val _group = getGroup(toCString(group.getName))
-    val err    = unistd.chown(toCString(path.toString), -1.toUInt, !(_group._2))
+  override def setGroup(group: GroupPrincipal): Unit =
+    Zone { implicit z =>
+      val _group = getGroup(toCString(group.getName))
+      val err    = unistd.chown(toCString(path.toString), -1.toUInt, !(_group._2))
 
-    if (err != 0) {
-      throw new IOException()
+      if (err != 0) {
+        throw new IOException()
+      }
     }
-  }
 
   override def readAttributes(): BasicFileAttributes =
     attributes
 
   private lazy val attributes =
     new PosixFileAttributes {
-      private val sb       = getStat()
-      private val mode     = !(sb._13)
-      private val _passwd  = getPasswd(!(sb._4))
-      private val _group   = getGroup(!(sb._5))
-      override val fileKey = (!(sb._3)).asInstanceOf[Object]
-
-      override val isDirectory    = stat.S_ISDIR(mode) == 1
-      override val isRegularFile  = stat.S_ISREG(mode) == 1
-      override val isSymbolicLink = stat.S_ISLNK(mode) == 1
-      override val isOther        = !isDirectory && !isRegularFile && !isSymbolicLink
-
-      override val lastAccessTime =
-        FileTime.from(!(sb._7), TimeUnit.SECONDS)
-      override val lastModifiedTime =
-        FileTime.from(!(sb._8), TimeUnit.SECONDS)
-      override val creationTime =
-        FileTime.from(!(sb._9), TimeUnit.SECONDS)
-
-      override val group = new GroupPrincipal {
-        override val getName = fromCString(!(_group._1))
+      private def fileStat()(implicit z: Zone) =
+        getStat()
```

Comment:
If you cache unmanaged pointer in a managed object, then you'll need to clean it up only when object itself is unreachable (i.e. garbaged collected.) On JVM you can use finalizers for that, but we don't support them at the moment, hence no caching.